### PR TITLE
[MCC][MCD]: Introduce in progress taint

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
**- What I did**
All the upgrade candidate nodes in a mcp would be applied
`UpdateInProgress: PreferNoSchedule` taint.
The taint will be removed MCD once the upgrade is complete.
Since kubernetes/kubernetes#104251 landed,
the nodes not having PreferNoSchedule taint will have higher score.
Before the upgrade starts, MCC will taint all the nodes in the
cluster that are supposed to be upgraded. Once the upgrade is
complete since MCD will remove the taint, none of the nodes will
have `UpdateInProgress: PreferNoSchedule` taint. This ensures
the score of the nodes will be equal again.

Why is this needed?
This reduces the pod churn when the cluster upgrade is in progress.
When the non-upgraded nodes in the cluster have `UpdateInProgress:
PreferNoSchedule` taint, they would get lesser score and the pods
would prefer to land onto untainted(upgraded) nodes there by
reducing the chances of landing onto an unupgraded node which
can cause one more reschedule
**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: pkg/controller/node/node_controller.go
Comment (wking) [inline pkg/controller/node/node_controller.go:991]: nit: `setUpdateInPrgressTaint` -> `setUpdateInProgressTaint`

File: pkg/controller/node/node_controller.go
Comment (wking) [inline pkg/controller/node/node_controller.go:979]: possibly pull the key out of `NodeUpdateTaint.Key` here and a few lines above, so you don't have to worry about matching spelling/casing/typos.

File: pkg/controller/node/node_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/node/node_controller.go:977]:
```suggestion
		ctrl.logPool(pool, "Applying UpdateInProgress taint to %s",  node.Name)
```

File: pkg/controller/node/node_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/node/node_controller.go:498]: prob need a nil check here?

**Review by kikisdeliveryservice — COMMENTED** [review]
quick pass, generally makes sense. will come back tomorrow to finish looking.

File: pkg/controller/node/node_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/node/node_controller.go:1007]: should this be nil or len(newNode.Spec.Taints) == 0 

(goes back to my q about whether this will always be nil or if it can be empty)

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:1332]: len(newNode.Spec.Taints) == 0 ? same q as above

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:1308]: nice <3

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:1335]: I think I'm tired but can we get a comment here on this loop thru append are doing bc future me will also be tired and easily confused ;)

**Comment by kikisdeliveryservice** [comment]
Just for ref:

```
// Like TaintEffectNoSchedule, but the scheduler tries not to schedule
	// new pods onto the node, rather than prohibiting new pods from scheduling
	// onto the node entirely. Enforced by the scheduler.
TaintEffectPreferNoSchedule TaintEffect = "PreferNoSchedule"
``` 

https://github.com/kubernetes/kubernetes/blob/f0b7ad3ee06c5168fef5fa4f01fe445ece595f89/pkg/apis/core/types.go#L2684

File: pkg/controller/node/node_controller.go
Comment (kikisdeliveryservice) [inline pkg/controller/node/node_controller.go:74]: nit: when you consolidate these vars elsewhere, maybe add the "why" of this to the comment?

**Comment by sinnykumari** [comment]
Once this is ready, can we please add better description of why we are adding this taint in MCO. Also, will be nice to have details in commit message.

Also since feature freeze is on Friday, from the slack conversation I believe we leaned towards implementing this in 4.10.

File: pkg/controller/node/node_controller.go
Comment (ravisantoshgudimetla) [reply pkg/controller/node/node_controller.go:74]: Sure. Do you have a location where I can share code across MCO and MCC?

File: pkg/controller/node/node_controller.go
Comment (ravisantoshgudimetla) [reply pkg/controller/node/node_controller.go:498]: Why would one be non-nil empty and other be nil but I am if it is confusing, I can have `len()` checks as well.

File: pkg/controller/node/node_controller.go
Comment (ravisantoshgudimetla) [reply pkg/controller/node/node_controller.go:1007]: It depends on declaration. In this case, we have a slice but no backing array so nil is ok but there is nothing wrong in checking len as it works for both nil and empty slices. For example,

var a []string is a nil slice as it doesn't have underlying array
c := []string{} is an empty slice. 

Usually all the API objects returned will be nil slices as it encoding in JSON will be easier.

File: pkg/daemon/daemon.go
Comment (ravisantoshgudimetla) [reply pkg/daemon/daemon.go:1335]: I just want to remove a taint from the node. I can add a comment. Technically, I can convert this to set so that deletion will be easier but I don't think the current implementation is going to be that expensive considering the number of taints we have.

**Review by ravisantoshgudimetla — COMMENTED** [review]
Thank you for the reviews @sinnykumari @kikisdeliveryservice @wking 

> Once this is ready, can we please add better description of why we are adding this taint in MCO. Also, will be nice to have details in commit message.

Of course, I opened this to see if I am in the right direction or not.

>Also since feature freeze is on Friday, from the slack conversation I believe we leaned towards implementing this in 4.10.

I am fine with 4.10 implementation but I am wondering if this is too much of invasive change?

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [reply pkg/daemon/daemon.go:1335]: I think it's fine to leave implementation as-is now, just add comment :+1: 

we can refactor if necessary later

File: pkg/controller/node/node_controller.go
Comment (kikisdeliveryservice) [reply pkg/controller/node/node_controller.go:74]: i think we could maybe use constants.go?

https://github.com/openshift/machine-config-operator/blob/9c6c2bfd7ed498bfbc296d530d1839bd6a177b0b/pkg/constants/constants.go

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:11]: failing verify here: 

`pkg/daemon/daemon.go:11: File is not `goimports`-ed (goimports)`

File: pkg/controller/node/node_controller.go
Comment (soltysh) [reply pkg/controller/node/node_controller.go:498]: I doubt spec will ever be nil at this point in time, as for `Taints` `DeepEqual` handles nil case, for you.

File: pkg/controller/node/node_controller.go
Comment (soltysh) [inline pkg/controller/node/node_controller.go:75]: Nit: `NodeUpdateInProgressTaint`?

File: pkg/daemon/daemon.go
Comment (soltysh) [inline pkg/daemon/daemon.go:1343]: Alternatively, we can just iterate over taints leaving out the one we want to remove and then assign that. The complexity is similar, a bit more expensive memory wise, but easier to read, imo.

**Comment by sinnykumari** [comment]
> I am fine with 4.10 implementation but I am wondering if this is too much of invasive change?

I understand this change is not much invasive. Main hesitation is because of unplanned changes that get added in repo without having full context and time crunch makes things worse. With limited team capacity and already other planned work and unplanned bugs, it becomes a bit difficult. Having some time buffer allows to think and review code better for future maintenance. 

Also whatever code we write, we all do our best to test and review and think that this shouldn't break existing fetaures but we all know what happens in reality ;)

**Comment by kikisdeliveryservice** [comment]
> > I am fine with 4.10 implementation but I am wondering if this is too much of invasive change?
> 
> I understand this change is not much invasive. Main hesitation is because of unplanned changes that get added in repo without having full context and time crunch makes things worse. With limited team capacity and already other planned work and unplanned bugs, it becomes a bit difficult. Having some time buffer allows to think and review code better for future maintenance.
> 
> Also whatever code we write, we all do our best to test and review and think that this shouldn't break existing fetaures but we all know what happens in reality ;)

Agree with Sinny. next release is the better time frame. FTR I don't think this is a PR quality issue and more of a review/testing/discussion issue and I don't see a reason we can't work together to get this in next time (on the earlier side) pending those convos and testing/review/ :smile:

File: pkg/controller/node/node_controller.go
Comment (ravisantoshgudimetla) [reply pkg/controller/node/node_controller.go:498]: That is correct. Technically, the code above is right. I believe @kikisdeliveryservice's concern is more on the lines of readability.

**Comment by ravisantoshgudimetla** [comment]
https://github.com/kubernetes/kubernetes/pull/104251 merged upstream, once https://github.com/openshift/kubernetes/pull/1076 lands, we should be able to use beta3 version which has increased weights

**Comment by kikisdeliveryservice** [comment]
As per call, we'd also like an OTA approver on this PR along with a test verifying functionality before merging.

/assign @wking

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [reply pkg/daemon/daemon.go:11]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/2686/pull-ci-openshift-machine-config-operator-master-verify/1467926127505838080

**Comment by kikisdeliveryservice** [comment]
@ravisantoshgudimetla can you ping me for review once you've addressed the outstanding comments?

File: pkg/controller/node/node_controller.go
Comment (wking) [inline pkg/controller/node/node_controller.go]: you have a public `NodeUpdateBackoff` in `pkg/constants` now.  Can you drop this one?

File: pkg/controller/node/node_controller.go
Comment (wking) [inline pkg/controller/node/node_controller.go:967]: nit: the wrapper and log are both making some assumptions about what `setUpdateInProgressTaint` is up to.  Probably pretty safe assumptions, so feel free to ignore, but it would be slightly more robust to push those down into `setUpdateInProgressTaint` itself.

File: pkg/controller/node/node_controller.go
Comment (wking) [reply pkg/controller/node/node_controller.go]: fixed as of  652de25.

File: pkg/controller/node/node_controller.go
Comment (wking) [reply pkg/controller/node/node_controller.go]: and it looks like you do drop it in 652de252.  Maybe reshuffle your hunks between commits or squash down to one to avoid touching something in one commit and then removing it in the second?

File: pkg/controller/node/node_controller.go
Comment (wking) [inline pkg/controller/node/node_controller.go:985]: nit: can we add a context argument to `setUpdateInProgressTaint` to push this `context.TODO` up a level?  No need to plumb `Context` through legacy call stacks, but new functions that need a context should take a context argument to avoid growing even more tech debt ;).

File: pkg/daemon/daemon.go
Comment (wking) [inline pkg/daemon/daemon.go:1325]: nit: you can avoid a no-op patch by initializing to `-1` and then `return nil` if you're still `-1` after iterating through looking for a match.

**Review by wking — APPROVED** [review]
I've gone through this again and left a few local-scoped nits.  But nothing update-specific; everything in here is good on that front.  So feel free to address as many or few of my nits as you like for me, and once the MCO folks are happy, I have no problems with this landing however it ends up.

File: pkg/daemon/daemon.go
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:1332]: If `newNode.Spec.Taints = []corev1.Taint{}` still is the case here, wouldn't this panic?

Comment (wking) [reply pkg/daemon/daemon.go:1332]: [this][1] would be one way to avoid that panic.

[1]: https://github.com/openshift/machine-config-operator/pull/2686#discussion_r764383642

Comment (yuqi-zhang) [reply pkg/daemon/daemon.go:1332]: Ah right. Alternatively if the taint is not found, we could probably return early above.

File: pkg/controller/node/node_controller.go
Comment (ravisantoshgudimetla) [reply pkg/controller/node/node_controller.go:967]: I thought, we usually error or log but not both in fn or method. IIUC, you want log to be present in `setUpdateInProgressTaint` method?

File: pkg/daemon/daemon.go
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:1326]: instead of doing it this way maybe doing it the way soltyish suggests below is more readable/less confusing? this block feels a little hard to understand on a run thru

Comment (kikisdeliveryservice) [reply pkg/daemon/daemon.go:1326]: loop thru array, appending everything that isn't a taint, if len(clean_array) != len(original array) assign it to newNode.Spec.Taints else return nil

File: pkg/controller/node/node_controller.go
Comment (kikisdeliveryservice) [reply pkg/controller/node/node_controller.go:968]: what ravi did here seems consistent with the func call above

Comment (kikisdeliveryservice) [reply pkg/controller/node/node_controller.go:968]: tho @wking is there a specific change to the error message above you think would be better?

File: pkg/controller/node/node_controller.go
Comment (sinnykumari) [inline pkg/controller/node/node_controller.go:966]: s/Applying/applying/

File: pkg/controller/node/node_controller.go
Comment (sinnykumari) [inline pkg/controller/node/node_controller.go:967]: Is the goal is to apply UpdateInProgress taint only to nodes that are currently being updated?
I may have misunderstood but I thought we want to apply UpdateInProgress taint to all the nodes in the pool that has not been updated yet. And scheduler will make use of this taint to ensure that we try to rescheduled drained pods to node that doesn't have that taint(implies node has already been updated).

**Review by sinnykumari — COMMENTED** [review]
Thanks Ravi for updating the PR.
Added few inline comments. Additionally, an you update description of the PR and commit message with details on what this change is about. This will help anyone who is freshly looking at the PR.

File: pkg/controller/node/node_controller.go
Comment (ravisantoshgudimetla) [reply pkg/controller/node/node_controller.go:967]: I thought:

All the nodes in the pool that have not been updated yet == All the nodes that doesn't have desired configuration? Is it not the case?

File: pkg/controller/node/node_controller.go
Comment (ravisantoshgudimetla) [reply pkg/controller/node/node_controller.go:966]: I thought it is ok to have capital letters for log whereas it is recommended to not have error string starting with capitals.

File: pkg/controller/node/node_controller.go
Comment (sinnykumari) [reply pkg/controller/node/node_controller.go:967]: Desired Configuration are not set on all node together when a new update starts. Number of nodes where we set desiredConfig is based on maxUnavailable value set. So, by default it will be set only on one node that has been picked up for update. See https://github.com/openshift/machine-config-operator/blob/master/pkg/controller/node/node_controller.go#L770, we calculate number of candidates that can be updated before calling updateCandidateMachines()

**Comment by sinnykumari** [comment]
Thanks for updating the commit message and description.

**Review by sinnykumari — APPROVED** [review]
Overall LGTM. I don't see any major review comment left from others that can change behavior of this feature. Let's get this merged and if any minor concerns needs to get fixed, it can be fixed as follow-up PR.

Also, with Trevor's approval we should be good to go from upgrade team side.

/lgtm

File: pkg/controller/node/node_controller_test.go
Comment (ravisantoshgudimetla) [inline pkg/controller/node/node_controller_test.go:6]: oops ordering needs to be sorted out.

**Pull request merged**
